The following commit contains additions to dietpi-software for multip…

xenfomation opened the pull request:
The following pull request has several tickets-worth of updates.

363 - Stub | ViFM to Software Additional
Added option to install ViFM - a CLI-based file manager with Vi bindings under
Software Additional

357 - DietPi-Software | GNUStep Desktop Environment #357
Added option under Software Optimized to install GNUStep -
a lightweight, from-the-ground-up desktop environment.

347 - No Machine & stock configs
191 - DietPi-Software | Add xrdp installation option
Added an new Category under Software Optimized called Remote Desktop
I added the code necessary to pull No Machine's deb package and to run
this multi-platform, secure, very efficient Remote Desktop Server.
It also acts as a client - looking for hosts running No Machine servers
with port 4000 open

For XRDP, it simply installs and works.  Both packages follow previous
examples for ALSA, X components, and if no Desktop environment is selected,
the default goes to LXDE

195 - DietPi-Software | Add QuiteRSS installation option
I have added QuiteRSS under Desktops and Utilities (Software Additional)
I have also ensured that it install X dependencies, such as VNC Server does

151 - DietPi-Software | Add MATE desktop installation option
I added, under Software Additional as it is not optimized, the full MATE
Desktop environment with IceWeasel, etc.

If/Then conditions for Desktops/Software that require X, ALSA, and a
Desktop environment were added to the original section that checks
for these types of installations.

xenfomation committed:
The following commit contains additions to dietpi-software for multiple tickets.

363 - Stub | ViFM to Software Additional
Added option to install ViFM - a CLI-based file manager with Vi bindings under
Software Additional

357 - DietPi-Software | GNUStep Desktop Environment #357
Added option under Software Optimized to install GNUStep -
a lightweight, from-the-ground-up desktop environment.

347 - No Machine & stock configs
191 - DietPi-Software | Add xrdp installation option
Added an new Category under Software Optimized called Remote Desktop
I added the code necessary to pull No Machine's deb package and to run
this multi-platform, secure, very efficient Remote Desktop Server.
It also acts as a client - looking for hosts running No Machine servers
with port 4000 open

For XRDP, it simply installs and works.  Both packages follow previous
examples for ALSA, X components, and if no Desktop environment is selected,
the default goes to LXDE

195 - DietPi-Software | Add QuiteRSS installation option
I have added QuiteRSS under Desktops and Utilities (Software Additional)
I have also ensured that it install X dependencies, such as VNC Server does

151 - DietPi-Software | Add MATE desktop installation option
I added, under Software Additional as it is not optimized, the full MATE
Desktop environment with IceWeasel, etc.

If/Then conditions for Desktops/Software that require X, ALSA, and a
Desktop environment were added to the original section that checks
for these types of installations.

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -533,22 +540,135 @@ INDEX_WEBSERVER_TARGET=${INDEX_WEBSERVER_TARGET:=-2}
+			INSTALL_DESCRIPTION="NoMachine (Remote Desktop Server/Client)"
+			Banner_Installing
+
+			## HW MODELS NEED TO BE CONSIDERED
```
Yep :+1: Probably best to use the new `HW_ARCH` identifier: https://github.com/Fourdee/DietPi/blob/testing/dietpi/dietpi-obtain_hw_model#L39-L46

Example usage: https://github.com/Fourdee/DietPi/blob/testing/dietpi/dietpi-software#L1989-L2018

xenfomation replied:
THANK YOU!  I am working on this right now for #364 / #347 (NoMachine Architecture).

For reference - just for Linux - these are all the archs and OSes they support:
[https://www.nomachine.com/download/linux&id=1](https://www.nomachine.com/download/linux&id=1)

I am currently leveraging the HW_ARCH identifier to ensure the right package gets installed, or if HW_ARCH == 0, "Say sorry!"

Cheers!

Fourdee commented on dietpi/dietpi-autostart:
```diff
@@ -90,6 +90,8 @@ _EOF_
 			"4" "OpenTyrian"
 			"5" "DietPi-Cloudshell"
+			"6" "MATE: Desktop"
+			"7" "GNUStep: Desktop"
```
You'll also need to add code to run these desktops in the login script: https://github.com/Fourdee/DietPi/blob/testing/dietpi/login#L65-L101

xenfomation replied:
AH!  Thank you!  I had searched and searched for those as so startx, etc would get kicked off.  

**(I am book-marking everything!)**  

This item is corrected on my dev machine.

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -533,22 +540,135 @@ INDEX_WEBSERVER_TARGET=${INDEX_WEBSERVER_TARGET:=-2}
+		#Installs that require a DESKTOP
 		if (( $TIGHTVNCSERVER == 1 )); then
-			DESKTOP_LXDE=1
+			if (( $DESKTOP_LXDE == 0 )) && (( $DESKTOP_MATE == 0 )) && (( $DESKTOP_GNUSTEP == 0 )); then
```
I think we might need a "desktop preference system"? similar to the webserver one. I can add this in if you think its worth doing?

Also, when software is installed, the value is `2`, when software is "to be installed" (selected for install), the value is `1`. 
So when checking for both installed, and tobe installed, you can use `if (( $MYPROGRAM > 0 )); then`

xenfomation replied:
I agree about implementing the Desktop Preference system at some point, @Fourdee.

And @rhkean -- still no coffee yet :-)  I will discuss more on #41 how much I agree with the idea of implementing a Desktop Preference system.

To remain on point, I took your pointers and re-tested my code.  This came to a work around as my nest of if/then conditions was a bloody mess.  I simplified it and re-tested by selecting all sorts of packages that would depend on a Desktop Environment to be chosen.

During the install process, naturally not DESKTOP_LXDE, MATE, nor GNUStep were selected.  When it came time to look for a DESKTOP to install, I added a prompt -- since LXDE is the original light-weight Desktop using whiptail:

![desktop_warning](https://cloud.githubusercontent.com/assets/10519408/15636506/dbd42634-25ce-11e6-8c7e-b4c577d11622.PNG)

After pressing enter -- presto!  The installer continues with LXDE and all remaining GUI-Centric-Software (QuiteRSS, etc) without fault or being confused.

On reboot, I validated the results:

![post-workaround](https://cloud.githubusercontent.com/assets/10519408/15636510/01278f5c-25cf-11e6-8f31-5c4e9e4975d3.PNG)

So, this isn't permanent -- but I look forward to feedback on this for now as I continue along this commit/pull.

Fourdee commented:
Looks good @xenfomation :+1: 

I'll go through it and add any comments.

rhkean commented:
I think someone needs to take the coffee away from @xenfomation .... ROFL

Are you mainlining it this weekend?  :smiley_cat:

xenfomation commented:
@rhkean  Hehehehe -- I sure am using this weekend to code for DietPi & update personal things.  Three whole days off and I am soaking up the time to get things done!  As for coffee, my body calls that water as I've had insomnia since I was a wee lad.

@Fourdee -- please do keep me update via comments as I remembered the > 2, but for some reason -- I was just focused on zeroes and ones!  Do let me know if, currently, you need me to go in and clean-up X & Y areas.  

As for the Desktop Preference System: I thinks it's splendid.  I had a hard time justifying why I would put MATE under Software Optimized when it's nothing like GNUStep and other Desktops (fvwm, CDE, etc which are all slim).

I will go back to each section (on this page) and update what I have corrected/sync'd in with the right bits!

Cheers!

xenfomation commented:
As for NoMachine #347 -- I have added the logic based on HW_ARCH.  

As for MATE and GNUStep -- I have added the correct entries into login for autostart.

xenfomation committed:
Based on today's earlier commit - 364 - I have corrected the
following:

- Installations for XRDP, NoMachine, and additional software
  are set to a value of 2 on completion
- For NoMachine, I used the HW_ARCH value to determine which
  NoMachine package to download and install
- I added a work around where graphical programs depend on
  a DESKTOP selection being made.  If no selection is made,
  dietpi-software will notify the user that a desktop
  selection was expected & is needed: it defaults to LXDE.
- I updated login and autostart to properly leverage
  startx for Autologin related to MATE and GNUStep

Fourdee commented:
@xenfomation 
Thanks for the update and changes :+1:  I want to get v119 completed and released 1st, then i'll merge this in for v120. I'll also sort the merge conflicts out as I've made alot of sourcecode changes today.

xenfomation commented:
Sounds great to me -- just getting around to checking email and updates, so let me know if I can be of any help to you!

Fourdee commented:
@xenfomation 
Looks good :+1:  just going through it now to ensure merge conflict didnt miss anything.

Fourdee commented:
Ok, just making some notes of changes as I do them, will commit once completed:
- Moved MATE to dietpi list, rather than linux
- functioned creating dietpi desktop icons/apps
- Simplified the "installs that require a desktop code"
- Added uninstall code for mate/gnustep/nomachine/quiterss/xrdp in dietpi-uninstall
- corrected NoMachine download links. arm6 for RPi 1, arm7 for RPi 2/3
- Moved QUITERSS to Install_linux_Software()
- Shuffled the Desktop categories.
- Added AUTO_DietpiSoftware_Install_ entries in dietpi.txt
- Merged all desktop autostart options into one.

#### Testing required arm64 (C2):
- :u6307: Mate
- :u6307: GNUstep

# 
- :u6307: LXDE  + XRDP
- :u6307: Mate + VNC4 | https://github.com/Fourdee/DietPi/pull/364#issuecomment-223750138
- :u6307: : Mate + XRDP
- :u6307: GNUstep + VNC4 | https://github.com/Fourdee/DietPi/pull/364#issuecomment-223750138
- :u6307: GNUstep + XRDP

#### Testing required arm6+ :RPi:
- :u6307: Mate
- :u6307: GNUstep

# 
- :u6307: LXDE + NoMachine
- :u6307:  Mate + NoMachine
- :u6307: GNUstep + NoMachine

#### Documentation
- XRDP requires a running desktop (eg: autostart)

Fourdee commented:
@xenfomation 
XRDP login:
- Module?
- Username?
- Password?

Tried with root/dietpi sesman-Xvnc, mate installed:
![image](https://cloud.githubusercontent.com/assets/7077026/15781506/5165ed6a-299e-11e6-952f-df6c9addc50e.png)

Fourdee commented:
@xenfomation 
NoMachine is great, really impressed :+1: 
![image](https://cloud.githubusercontent.com/assets/7077026/15783834/be25476c-29a7-11e6-9eb5-277794c495d4.png)

xenfomation commented:
Awesome on NoMachine!!

### XRDP

Ouch -- glad you raised that as my first test last month before coding the request worked spiffy.

I hope this is not the cause: [https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=776746](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=776746)

Was your X environment started by default or started by running startx?

I will re-test again w/ Mate and have results for you ASAP.  I am working off of v119 on VMs to save on SD wear and tear until I bag a few more this weekend.

Thanks, @Fourdee !

xenfomation commented:
Okay - I was able to successfully login via XRDP from a Windows Host to DietPi v119 w/ Mate, Xsession/Xauth, etc installed.

Pardon the screenshots, but they are in order of operations from Clean DietPi v119 -- I only installed OpenSSH Client/Server, Xsession/Xauth, Mate, and XRDP.

Screenshot 12 shows the "Inception Moment" where my v119 DietPi w/ MATE installed is in the background of using Remote Desktop Connection to it :-)

![xrdp1](https://cloud.githubusercontent.com/assets/10519408/15797609/bf2955d2-29ec-11e6-91f5-af597769d635.PNG)
![xrdp2](https://cloud.githubusercontent.com/assets/10519408/15797612/bf30c7f4-29ec-11e6-97f3-893586291cc8.PNG)
![xrdp3](https://cloud.githubusercontent.com/assets/10519408/15797613/bf31bde4-29ec-11e6-8457-6d7fbf627227.PNG)
![xrdp4](https://cloud.githubusercontent.com/assets/10519408/15797610/bf2d787e-29ec-11e6-8455-0d2aa33bd7c3.PNG)
![xrdp5](https://cloud.githubusercontent.com/assets/10519408/15797615/bf31fd90-29ec-11e6-8d14-e5dc4f69bec7.PNG)
![xrdp6](https://cloud.githubusercontent.com/assets/10519408/15797611/bf2e569a-29ec-11e6-8712-d5f12f008aae.PNG)
![xrdp7](https://cloud.githubusercontent.com/assets/10519408/15797614/bf320010-29ec-11e6-8766-970ac64737a0.PNG)
![xrdp8](https://cloud.githubusercontent.com/assets/10519408/15797616/bf371794-29ec-11e6-96d7-88c02480f516.PNG)
![xrdp9](https://cloud.githubusercontent.com/assets/10519408/15797617/bf37f61e-29ec-11e6-98a1-443e9248398f.PNG)
![xrdp10](https://cloud.githubusercontent.com/assets/10519408/15797618/bf3b168c-29ec-11e6-84c1-d9725ddcc3f6.PNG)
![xrdp11](https://cloud.githubusercontent.com/assets/10519408/15797619/bf3b79d8-29ec-11e6-94df-70132921a8fa.PNG)
![xrdp12](https://cloud.githubusercontent.com/assets/10519408/15797620/bf3edfb0-29ec-11e6-8fc4-7e9645601e52.PNG)

Fourdee commented:
@xenfomation 

Thanks. So XRDP requires a running desktop. Unlike VNC which can start the desktop automatically.

Yep, working fine :+1: :
![image](https://cloud.githubusercontent.com/assets/7077026/15799124/2d9c98aa-2a4a-11e6-8353-c718fa1ba6d1.png)

Fourdee commented:
C2 mate/gnustep desktop + VNC4

```
/root/.vnc/xstartup: 5: /root/.vnc/xstartup: /usr/bin/lxsession: not found
```

Fixed:

```
        #VNC4SERVER
        if (( $VNC4SERVER == 1 )); then


            local cmd_launch_desktop=''
            if (( $DESKTOP_LXDE > 0 )); then

                cmd_launch_desktop='/usr/bin/lxsession -s LXDE &'

            elif (( $DESKTOP_MATE > 0 )); then

                cmd_launch_desktop='x-window-manager &'

            elif (( $DESKTOP_GNUSTEP > 0 )); then

                cmd_launch_desktop='x-window-manager &'

            fi

            cat << _EOF_ > "$HOME"/.vnc/xstartup
[ -x /etc/vnc/xstartup ] && exec /etc/vnc/xstartup
[ -r /root/.Xresources ] && xrdb /root/.Xresources
xsetroot -solid grey
vncconfig -iconic &
$cmd_launch_desktop
_EOF_
```

Tested GNUstep with tightvnc. Worked straight away and doesnt require the vnc4 changes above.

xenfomation commented:
Sweet and nicely caught, Sir!  If I can contribute to any of the docs, let me know as I am working on the XenServer Virtual Machine Image (XVA format) documentation.
